nixos/modules/services/cluster/k3s/default.nix

```diff
@@ -434,6 +434,25 @@ in
         for further information.
       '';
     };
+
+    extraKubeProxyConfig = lib.mkOption {
+      type = with lib.types; attrsOf anything;
+      default = {
+        clientConnection = {
+          kubeconfig = "/etc/rancher/k3s/k3s.yaml";
+        };
+      };
```

Comment:
I'm not sure whether I get your comments right, but I think this is a wrong default. K3s actually doesn't use this kubeconfig for kube-proxy. By default it passes kube-proxy the argument `--kubeconfig=/var/lib/rancher/k3s/agent/kubeproxy.kubeconfig`, which is another file. I suggest to make the default `{}`, so that the `--kube-proxy-arg=config=...` gets only passed if a user explicitly changes this option.

Reply:
k3s will pass `kubeconfig` to kube-proxy in https://github.com/k3s-io/k3s/blob/4ec261733e241e2baf77684ad5d87ff071d0f00a/pkg/daemons/agent/agent_linux.go#L43 via command line arg `--kube-config` , then kube-proxy will save this arg to `o.config.ClientConnection.Kubeconfig` in https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-proxy/app/options.go#L106 . However, the `o.config` object will be override in https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-proxy/app/options.go#L221 by the kubeProxyConfigurataion object which specified via the `config=path` option.

Reply:
> k3s will pass kubeconfig to kube-proxy in https://github.com/k3s-io/k3s/blob/4ec261733e241e2baf77684ad5d87ff071d0f00a/pkg/daemons/agent/agent_linux.go#L43 via command line arg --kube-config

I agree, but notice that the variable used here is `cfg.KubeConfigKubeProxy`, which gets constructed [here](https://github.com/k3s-io/k3s/blob/4ec261733e241e2baf77684ad5d87ff071d0f00a/pkg/agent/config/config.go#L518) . K3s generates multiple kubeconfigs with different certificates for different components, e.g. kubelet, kubeconfig, k3s-controller, etc. I don't know the internals but we shouldn't change which kubeconfig is used without good reason. The default kubeconfig for kube-proxy is at `/var/lib/rancher/k3s/agent/kubeproxy.kubeconfig` and not the admin kubeconfig at `/etc/rancher/k3s/k3s.yaml`. 

You can see which args are passed to kube-proxy with `journalctl -u k3s | grep kube-proxy`, on the current master **without** this PR and default settings I see:

`k3s[778]: time="2024-12-08T14:05:20Z" level=info msg="Running kube-proxy --cluster-cidr=10.42.0.0/16 --conntrack-max-per-core=0 --conntrack-tcp-timeout-close-wait=0s --conntrack-tcp-timeout-established=0s --healthz-bind-address=127.0.0.1 --hostname-override=foo --kubeconfig=/var/lib/rancher/k3s/agent/kubeproxy.kubeconfig --proxy-mode=iptables"`

The actual default for `extraKubeProxyConfig` would change the kubeconfig that kube-proxy uses. Of course we could change the default, but this would lead to other problems, for example if k3s changes its defaults or the k3s data dir is changed. I think it would be best to not generate a config for kube-proxy by default, so that k3s can decide which kubeconfig it passes to kube-proxy. However, if somebody configures a custom `extraKubeProxyConfig`, he has to take care of setting `clientConnection.kubeconfig` as mentioned in the option description.

Reply:
> The actual default for `extraKubeProxyConfig` would change the kubeconfig that kube-proxy uses. Of course we could change the default, but this would lead to other problems, for example if k3s changes its defaults or the k3s data dir is changed. I think it would be best to not generate a config for kube-proxy by default, so that k3s can decide which kubeconfig it passes to kube-proxy. However, if somebody configures a custom `extraKubeProxyConfig`, he has to take care of setting `clientConnection.kubeconfig` as mentioned in the option description.

I agree that.